### ISSUE TITLE
extra tokens are not optimised by CTC loss but used as targets for model to predict

### ISSUE THREAD

**Issue by wenjie-p** [s3prl/s3prl#570]
Hi @leo19941227 ,

I have some questions about the models optimised by CTC loss. I found the extra tokens like `<pad>` and `<eos>` are used as the targets for model to predict. Since these tokens are not required for ctc loss calculation, they will not be optimised. Therefore, my question is whether it is necessary to include these tokens as targets for the model to predict.

For the downstream `ctc`, the output units include `self._vocab_list = ["<pad>", "<eos>", "<unk>"] + vocab_list` , as shown in this [line](https://github.com/s3prl/s3prl/blob/main/s3prl/downstream/ctc/text.py#L99). The `<pad>` and `<eos>` are excluded for the calculation of CTC loss and PER as shown in this code block

```
      loss = self.objective(
            log_probs.transpose(0, 1),  # (N, T, C) -> (T, N, C)
            labels,
            log_probs_len,
            labels_len,
        )
        records["loss"].append(loss.item())
        pred_tokens = log_probs.argmax(dim=-1)
        filtered_tokens = []
        for pred_token in pred_tokens:
            pred_token = pred_token.unique_consecutive()
            filtered_token = [
                token
                for token in pred_token.tolist()
                if token != self.tokenizer.pad_idx and token != self.tokenizer.eos_idx
            ]
```

This is similar to the downstream `asr`.  I trained PR, ASR and SF using the last layer of wavlm base, and found exclude the extra tokens as prediction targets can improve a bit on PR and ASR, but in SF it improves on the slot-value cer but degrades slightly on slot-type-f1 on test set.

Below are my numbers:

- PR
    - w/ extra tokens: dev per: 5.19, test per: 5.60
    - w/o extra tokens: dev per: 5.12 test per: 5.47
- ASR
    - w/ extra tokens: dev wer 5.80, test wer: 6.39
    - w/o extra tokens: dev wer 5.69, test wer: 6.23
- SF
    - w/ extra tokens dev: slot-type-f1: 88.40; slot-value-cer: 26.95, test: slot-type-f1: 87.10, slot-value-cer: 27.11
    - w/o extra tokens dev: slot-type-f1: 88.87; slot-value-cer: 26.50, test: slot-type-f1: 86.99, slot-value-cer: 26.65
